From 69f365b487ec14637ee468c3d61f7d562e850f4f Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Tue, 13 Feb 2018 09:33:46 -0800
Subject: [PATCH 0274/1852] drm: xlnx: pl_disp: Use xlnx pipeline calls

commit dd254f165495e6cacff9e4151be1dec0299d2aca from
https://github.com/Xilinx/linux-xlnx.git

Use xlnx_drm_pipeline_init() call as the dt binding is removed
for xlnx drm. The xlnx drm is more or less a library, and
should be explicitly initialized / released.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Saurabh Sengar <saurabhs@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_pl_disp.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_pl_disp.c b/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
index e2d6d752afda..66ee8477c26d 100644
--- a/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
+++ b/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
@@ -23,6 +23,7 @@
 #include <linux/of.h>
 #include <linux/of_dma.h>
 #include <linux/platform_device.h>
+#include "xlnx_drv.h"
 #include "xlnx_crtc.h"
 
 /*
@@ -50,6 +51,7 @@ struct xlnx_dma_chan {
 /**
  * struct xlnx_pl_disp - struct for display subsystem
  * @dev: device structure
+ * @master: logical master device from xlnx drm
  * @xlnx_crtc: Xilinx DRM driver crtc object
  * @plane: base drm plane object
  * @chan: struct for DMA engine
@@ -61,6 +63,7 @@ struct xlnx_dma_chan {
  */
 struct xlnx_pl_disp {
 	struct device *dev;
+	struct platform_device *master;
 	struct xlnx_crtc xlnx_crtc;
 	struct drm_plane plane;
 	struct xlnx_dma_chan *chan;
@@ -468,10 +471,19 @@ static int xlnx_pl_disp_probe(struct platform_device *pdev)
 	if (ret)
 		goto err_dma;
 
+	xlnx_pl_disp->master = xlnx_drm_pipeline_init(pdev);
+	if (IS_ERR(xlnx_pl_disp->master)) {
+		ret = PTR_ERR(xlnx_pl_disp->dev);
+		dev_err(dev, "failed to initialize the drm pipeline\n");
+		goto err_component;
+	}
+
 	dev_info(&pdev->dev, "Xlnx PL display driver probed\n");
 
 	return 0;
 
+err_component:
+	component_del(dev, &xlnx_pl_disp_component_ops);
 err_dma:
 	dma_release_channel(xlnx_pl_disp->chan->dma_chan);
 
@@ -483,6 +495,7 @@ static int xlnx_pl_disp_remove(struct platform_device *pdev)
 	struct xlnx_pl_disp *xlnx_pl_disp = platform_get_drvdata(pdev);
 	struct xlnx_dma_chan *xlnx_dma_chan = xlnx_pl_disp->chan;
 
+	xlnx_drm_pipeline_exit(xlnx_pl_disp->master);
 	component_del(&pdev->dev, &xlnx_pl_disp_component_ops);
 
 	/* Make sure the channel is terminated before release */
-- 
2.31.1

